# Signed Graph Convolutional Network(SGCN)を用いたFraud User Detection

```sh
pip install -r /home/ubuntu/SGCN/requirements.txt
pip install torch_scatter
pip install torch_sparse
pip install easydict
```


In [1]:
from sgcn import SignedGCNTrainer, SignedGCNPredictor
from parser import parameter_parser
from utils import tab_printer, read_graph, score_printer, save_logs
import easydict
import argparse
import pandas as pd
import numpy as np
import torch
import json
import networkx as nx
from sklearn.metrics import roc_auc_score, roc_curve, auc
from sklearn.metrics import accuracy_score,confusion_matrix
from IPython.display import display
import matplotlib.pyplot as plt
%matplotlib inline

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/matplotlib/__init__.py:1067: UserWarning: Duplicate key in file "/home/ubuntu/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/matplotlib/__init__.py:1067: UserWarning: Duplicate key in file "/home/ubuntu/.config/matplotlib/matplotlibrc", line #3
  (fname, cnt))


In [2]:
data_name = input('データセット：')

データセット：amazon


In [3]:
args = easydict.EasyDict({
        "edge_path": f'../input/{data_name}/{data_name}_network.csv',#'../input/{data_name}/user_network.csv',
        "features_path":  f'../input/{data_name}/{data_name}_node_feature.csv',#'../input/{data_name}/user_network.csv',
        "nodes_path": f'../input/{data_name}/{data_name}_gt.csv',
        "embedding_path": f'../tmp/embedding/{data_name}_sgcn_feature05.pkl', # tmp folder for cross-validation
        "regression_weights_path": f'../tmp/weights/{data_name}_sgcn_feature05.pkl',
        "inductive_model_path": None, # f'../output/inductive/{data_name}_model', # or None
        "log_path": f'../logs/{data_name}_logs_feature05.json',
        "epochs":300,
        "test_size":0.33,
        "reduction_iterations": 128,
        "reduction_dimensions": 30,
        "seed": 42,
        "lamb": 1.0,
        "learning_rate": 0.005,  
        "weight_decay": 10e-4, 
        # "layers": [64, 32,16,8],
        "layers": [32, 16, ],
        "spectral_features":False,
        "general_features": True,  
        "sample_num":None,
        "class_weights":False,
        "node_under_sampling":False,
        "hidden_residual":False,
        "eval_freq":1,
})

In [4]:
display(pd.io.json.json_normalize(args).T)

0
class_weights                                                 False
edge_path                        ../input/amazon/amazon_network.csv
embedding_path           ../tmp/embedding/amazon_sgcn_feature05.pkl
epochs                                                          300
eval_freq                                                         1
features_path               ../input/amazon/amazon_node_feature.csv
general_features                                               True
hidden_residual                                               False
inductive_model_path                                           None
lamb                                                              1
layers                                                  [32, 16, 8]
learning_rate                                                 0.005
log_path                         ../logs/amazon_logs_feature05.json
node_under_sampling                                           False
nodes_path                            ../input/amazon/amazon_gt.csv
reduction_dimensions                                             30
reduction_iterations                                            128
regression_weights_path    ../tmp/weights/amazon_sgcn_feature05.pkl
sample_num                                                     None
seed                                                             42
spectral_features                                             False
test_size                                                      0.33
weight_decay                                                  0.001

## 10-fold cross-validation
- train : validation : test = 6:3:1

In [5]:
#tab_printer(args)
edges, nodes_dict = read_graph(args) # nodes_dict['indice']:node_id , nodes_dict['label'] : label

In [6]:
from sklearn.model_selection import StratifiedKFold

In [7]:
kf = StratifiedKFold(n_splits=10)
all_indice = nodes_dict['indice']
all_labels = nodes_dict['label']
auc_scores = []

In [8]:
for i, (train_index, test_index) in enumerate(kf.split(X=nodes_dict['indice'],y=nodes_dict['label'])):
    print("==== Training Phase ====")
    print(f'{i}-th fold')
    # training
    train_node_indice = all_indice[train_index]
    train_node_labels = all_labels[train_index]
    print(f'labels:{np.unique(train_node_labels,return_counts=True)}')
    tmp_nodes_dict = {}
    tmp_nodes_dict['all_ncount'] = nodes_dict['all_ncount']
    tmp_nodes_dict['indice'] = train_node_indice
    tmp_nodes_dict['label'] = train_node_labels
    trainer = SignedGCNTrainer(args, edges, tmp_nodes_dict)
    trainer.setup_dataset()
    trainer.create_and_train_model()
    
    if args.test_size > 0:
        # trainer.save_model() ## trainer.create_and_train_model()のなかで，すでにbest_modelが保存されている．
        # score_printer(trainer.logs)
        display(pd.DataFrame(trainer.logs['performance']))
        save_logs(args, trainer.logs)

    # test
    print("==== Test Phase ====")
    test_node_indice = all_indice[test_index]
    test_node_labels = all_labels[test_index]
    # feature = pd.read_csv(args.embedding_path,index_col='id').values
    feature = pd.read_pickle(args.embedding_path).drop('id',1).values
    test_feature = feature[test_node_indice]
    # weight = pd.read_csv(args.regression_weights_path)
    weight = pd.read_pickle(args.regression_weights_path)
    predictions = np.dot(test_feature,weight.values.T)
    probabilities = torch.nn.functional.softmax(torch.from_numpy(predictions)).numpy()
    predict_labels = probabilities.argmax(1)
    auc_score = roc_auc_score(y_true=[1 if i==-1 else 0 for i in test_node_labels],y_score=probabilities[:,1])
    auc_scores.append(auc_score)
    cmx = confusion_matrix(y_true=[1 if i==-1 else 0 for i in test_node_labels],y_pred=predict_labels)
    print(f"{i}-th fold's auc_score:{auc_score}")
    print(cmx)
    print()
    

==== Training Phase ====
0-th fold
labels:(array([-1,  1]), array([ 216, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.183): 100%|██████████| 300/300 [08:27<00:00,  1.72s/it] 


0         1   2
0    Epoch       AUC  F1
1        1  0.293936   0
2        2  0.213448   0
3        3  0.200589   0
4        4  0.201533   0
5        5  0.217426   0
6        6  0.224157   0
7        7  0.227954   0
8        8   0.23798   0
9        9  0.250678   0
10      10  0.268902   0
11      11  0.284513   0
12      12   0.30565   0
13      13  0.330624   0
14      14  0.357186   0
15      15  0.393512   0
16      16  0.451257   0
17      17  0.526994   0
18      18  0.593508   0
19      19  0.645707   0
20      20  0.702467   0
21      21   0.73368   0
22      22  0.738954   0
23      23  0.741044   0
24      24  0.748177   0
25      25  0.754978   0
26      26  0.760172   0
27      27  0.765014   0
28      28  0.769032   0
29      29   0.77287   0
..     ...       ...  ..
271    271   0.85798   0
272    272  0.858381   0
273    273  0.859105   0
274    274  0.859426   0
275    275  0.860129   0
276    276  0.861466   0
277    277  0.861134   0
278    278  0.861496   0
279    279  0.862119   0
280    280  0.862641   0
281    281  0.863163   0
282    282  0.862058   0
283    283   0.86234   0
284    284  0.859567   0
285    285  0.858984   0
286    286  0.859607   0
287    287  0.857437   0
288    288  0.860993   0
289    289  0.861214   0
290    290  0.860823   0
291    291  0.861275   0
292    292  0.862199   0
293    293  0.862782   0
294    294  0.863485   0
295    295  0.863887   0
296    296  0.865373   0
297    297  0.865012   0
298    298   0.86457   0
299    299  0.865333   0
300    300  0.865876   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
0-th fold's auc_score:0.8145762711864406
[[225  11]
 [ 20   5]]

==== Training Phase ====
1-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.2155): 100%|██████████| 300/300 [08:27<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.325337   0
2        2  0.206726   0
3        3  0.184246   0
4        4  0.177937   0
5        5  0.190933   0
6        6  0.203353   0
7        7  0.213135   0
8        8  0.225853   0
9        9   0.23869   0
10      10  0.254524   0
11      11  0.272083   0
12      12      0.29   0
13      13  0.311667   0
14      14  0.338016   0
15      15  0.369147   0
16      16  0.410774   0
17      17  0.464425   0
18      18  0.522738   0
19      19  0.579425   0
20      20  0.627758   0
21      21  0.685476   0
22      22  0.729306   0
23      23  0.752877   0
24      24   0.76123   0
25      25  0.767857   0
26      26  0.775575   0
27      27  0.783175   0
28      28  0.793552   0
29      29   0.79998   0
..     ...       ...  ..
271    271  0.890635   0
272    272  0.890873   0
273    273  0.890556   0
274    274   0.89129   0
275    275  0.891587   0
276    276  0.891687   0
277    277  0.890615   0
278    278  0.890476   0
279    279   0.89127   0
280    280  0.891508   0
281    281  0.892183   0
282    282  0.891806   0
283    283  0.891865   0
284    284  0.891706   0
285    285  0.890774   0
286    286  0.892103   0
287    287  0.891647   0
288    288   0.89006   0
289    289   0.89246   0
290    290  0.890913   0
291    291   0.89127   0
292    292  0.891071   0
293    293  0.890754   0
294    294  0.891567   0
295    295  0.890516   0
296    296  0.893591   0
297    297  0.877937   0
298    298  0.892302   0
299    299  0.887877   0
300    300  0.890595   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
1-th fold's auc_score:0.9143714689265536
[[233   3]
 [ 17   7]]

==== Training Phase ====
2-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.2361): 100%|██████████| 300/300 [08:24<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.269028   0
2        2  0.182103   0
3        3     0.175   0
4        4  0.166548   0
5        5  0.179147   0
6        6   0.19871   0
7        7  0.210913   0
8        8  0.222103   0
9        9  0.240853   0
10      10  0.259563   0
11      11  0.282183   0
12      12  0.303502   0
13      13  0.331885   0
14      14   0.36375   0
15      15  0.400595   0
16      16  0.447163   0
17      17  0.516071   0
18      18  0.597996   0
19      19  0.654464   0
20      20  0.718214   0
21      21  0.765496   0
22      22  0.775933   0
23      23  0.781091   0
24      24  0.785972   0
25      25   0.79244   0
26      26  0.799901   0
27      27  0.807143   0
28      28  0.812639   0
29      29  0.815794   0
..     ...       ...  ..
271    271  0.879643   0
272    272    0.8775   0
273    273  0.876766   0
274    274  0.878056   0
275    275  0.876528   0
276    276  0.877302   0
277    277  0.873988   0
278    278  0.875694   0
279    279  0.877758   0
280    280  0.880357   0
281    281  0.878492   0
282    282  0.875218   0
283    283  0.877123   0
284    284  0.880476   0
285    285  0.879881   0
286    286  0.879385   0
287    287  0.875952   0
288    288  0.876825   0
289    289  0.876468   0
290    290  0.876389   0
291    291  0.878294   0
292    292  0.875992   0
293    293  0.876071   0
294    294  0.877738   0
295    295  0.878254   0
296    296  0.877837   0
297    297  0.879504   0
298    298  0.879802   0
299    299  0.868095   0
300    300  0.867599   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
2-th fold's auc_score:0.8774717514124294
[[230   6]
 [ 21   3]]

==== Training Phase ====
3-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.174): 100%|██████████| 300/300 [08:23<00:00,  1.65s/it] 


0         1   2
0    Epoch       AUC  F1
1        1  0.353284   0
2        2  0.257827   0
3        3  0.242986   0
4        4  0.233105   0
5        5  0.240169   0
6        6  0.248026   0
7        7  0.258006   0
8        8  0.266478   0
9        9  0.280565   0
10      10  0.297708   0
11      11  0.313145   0
12      12  0.334524   0
13      13  0.358264   0
14      14   0.38752   0
15      15  0.420556   0
16      16  0.474107   0
17      17  0.535437   0
18      18  0.591637   0
19      19  0.637857   0
20      20  0.670308   0
21      21  0.704613   0
22      22   0.71501   0
23      23  0.719355   0
24      24  0.722966   0
25      25  0.728353   0
26      26  0.732728   0
27      27  0.738899   0
28      28  0.745407   0
29      29  0.751617   0
..     ...       ...  ..
271    271  0.838542   0
272    272  0.839752   0
273    273  0.839375   0
274    274  0.838839   0
275    275  0.843403   0
276    276  0.840466   0
277    277  0.839871   0
278    278  0.839712   0
279    279  0.837808   0
280    280  0.844534   0
281    281    0.8387   0
282    282  0.841915   0
283    283  0.836796   0
284    284  0.837589   0
285    285  0.841796   0
286    286  0.839196   0
287    287    0.8413   0
288    288  0.836954   0
289    289  0.839474   0
290    290  0.838661   0
291    291  0.838978   0
292    292  0.840288   0
293    293  0.838462   0
294    294  0.842054   0
295    295  0.838046   0
296    296  0.842867   0
297    297  0.835466   0
298    298  0.839137   0
299    299  0.836478   0
300    300  0.833661   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
3-th fold's auc_score:0.8200918079096046
[[231   5]
 [ 20   4]]

==== Training Phase ====
4-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.1898): 100%|██████████| 300/300 [08:20<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.324802   0
2        2  0.219802   0
3        3  0.202837   0
4        4  0.200694   0
5        5  0.220099   0
6        6  0.226091   0
7        7   0.23881   0
8        8  0.254425   0
9        9  0.270159   0
10      10  0.285694   0
11      11  0.304345   0
12      12  0.326587   0
13      13  0.355099   0
14      14  0.387381   0
15      15  0.435595   0
16      16  0.513056   0
17      17  0.591806   0
18      18  0.643869   0
19      19  0.680675   0
20      20  0.715754   0
21      21   0.74502   0
22      22  0.753036   0
23      23   0.75996   0
24      24  0.766587   0
25      25  0.773988   0
26      26  0.784087   0
27      27  0.795179   0
28      28  0.806806   0
29      29  0.814375   0
..     ...       ...  ..
271    271  0.873968   0
272    272  0.876925   0
273    273  0.875913   0
274    274  0.873373   0
275    275  0.872282   0
276    276  0.874722   0
277    277  0.874643   0
278    278  0.872837   0
279    279  0.872758   0
280    280  0.873968   0
281    281  0.873413   0
282    282  0.872004   0
283    283  0.873452   0
284    284  0.871706   0
285    285  0.866389   0
286    286  0.865774   0
287    287  0.866429   0
288    288  0.865635   0
289    289  0.865198   0
290    290  0.866925   0
291    291   0.86631   0
292    292  0.867163   0
293    293  0.865655   0
294    294  0.864583   0
295    295  0.866607   0
296    296  0.865833   0
297    297  0.866825   0
298    298  0.865377   0
299    299  0.866329   0
300    300  0.863869   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
4-th fold's auc_score:0.7547669491525424
[[229   7]
 [ 20   4]]

==== Training Phase ====
5-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.1942): 100%|██████████| 300/300 [08:20<00:00,  1.66s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.365139   0
2        2   0.24873   0
3        3  0.223948   0
4        4  0.214266   0
5        5  0.222857   0
6        6  0.227044   0
7        7  0.231687   0
8        8  0.241329   0
9        9  0.249881   0
10      10  0.264306   0
11      11  0.278155   0
12      12   0.29881   0
13      13  0.323433   0
14      14   0.34877   0
15      15  0.382103   0
16      16  0.436647   0
17      17  0.501111   0
18      18  0.547897   0
19      19  0.589732   0
20      20  0.641369   0
21      21  0.678254   0
22      22  0.690575   0
23      23  0.698968   0
24      24  0.706687   0
25      25  0.712123   0
26      26  0.719772   0
27      27  0.726329   0
28      28   0.73248   0
29      29  0.742014   0
..     ...       ...  ..
271    271  0.836012   0
272    272  0.832282   0
273    273  0.831548   0
274    274  0.829563   0
275    275  0.828909   0
276    276  0.829087   0
277    277  0.830595   0
278    278  0.831806   0
279    279  0.832976   0
280    280  0.835238   0
281    281   0.83494   0
282    282  0.834187   0
283    283  0.835456   0
284    284  0.835575   0
285    285  0.833274   0
286    286   0.83256   0
287    287  0.831875   0
288    288  0.834702   0
289    289  0.835704   0
290    290  0.838512   0
291    291  0.843452   0
292    292  0.845159   0
293    293  0.847222   0
294    294  0.844782   0
295    295  0.833492   0
296    296  0.823631   0
297    297  0.815417   0
298    298  0.824385   0
299    299  0.830714   0
300    300  0.836825   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
5-th fold's auc_score:0.9336158192090396
[[235   1]
 [ 15   9]]

==== Training Phase ====
6-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.2065): 100%|██████████| 300/300 [08:26<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.355486   0
2        2  0.225685   0
3        3  0.218978   0
4        4  0.220387   0
5        5  0.238065   0
6        6  0.248998   0
7        7  0.259038   0
8        8  0.268284   0
9        9  0.280883   0
10      10  0.294772   0
11      11   0.31249   0
12      12  0.330367   0
13      13  0.352927   0
14      14  0.379038   0
15      15  0.408403   0
16      16  0.451815   0
17      17  0.508919   0
18      18   0.57379   0
19      19  0.626657   0
20      20  0.673492   0
21      21   0.71749   0
22      22  0.734097   0
23      23  0.740179   0
24      24  0.747133   0
25      25  0.755069   0
26      26  0.763006   0
27      27   0.77378   0
28      28  0.783343   0
29      29  0.790546   0
..     ...       ...  ..
271    271  0.858998   0
272    272  0.858899   0
273    273  0.858462   0
274    274  0.858462   0
275    275  0.858185   0
276    276  0.858383   0
277    277  0.859276   0
278    278  0.858859   0
279    279  0.858661   0
280    280  0.857887   0
281    281  0.858601   0
282    282  0.859077   0
283    283  0.858542   0
284    284  0.858383   0
285    285  0.858978   0
286    286  0.858819   0
287    287  0.858879   0
288    288  0.858978   0
289    289  0.858274   0
290    290  0.858859   0
291    291  0.859038   0
292    292  0.858978   0
293    293  0.858482   0
294    294  0.858542   0
295    295    0.8588   0
296    296  0.857728   0
297    297  0.857986   0
298    298  0.858284   0
299    299  0.857768   0
300    300  0.857649   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
6-th fold's auc_score:0.891949152542373
[[236   0]
 [ 24   0]]

==== Training Phase ====
7-th fold
labels:(array([-1,  1]), array([ 217, 2122]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.201): 100%|██████████| 300/300 [08:24<00:00,  1.65s/it] 


0         1   2
0    Epoch       AUC  F1
1        1  0.321419   0
2        2  0.204613   0
3        3  0.192391   0
4        4  0.184851   0
5        5  0.200446   0
6        6  0.212093   0
7        7  0.221181   0
8        8  0.240655   0
9        9  0.259812   0
10      10  0.276577   0
11      11  0.304157   0
12      12  0.325724   0
13      13  0.349573   0
14      14  0.377569   0
15      15  0.411657   0
16      16  0.464931   0
17      17  0.532014   0
18      18  0.588879   0
19      19  0.642391   0
20      20  0.705774   0
21      21  0.748026   0
22      22  0.764534   0
23      23  0.772252   0
24      24  0.778641   0
25      25  0.785724   0
26      26  0.792946   0
27      27   0.79999   0
28      28  0.807192   0
29      29  0.814415   0
..     ...       ...  ..
271    271  0.901379   0
272    272  0.902867   0
273    273  0.902034   0
274    274  0.901657   0
275    275  0.901935   0
276    276  0.903363   0
277    277  0.904117   0
278    278  0.903323   0
279    279  0.901895   0
280    280  0.903562   0
281    281  0.902232   0
282    282  0.905446   0
283    283  0.898681   0
284    284  0.905367   0
285    285  0.896101   0
286    286  0.902312   0
287    287  0.902986   0
288    288  0.901657   0
289    289  0.898204   0
290    290  0.898899   0
291    291  0.899534   0
292    292  0.900149   0
293    293   0.90378   0
294    294  0.901359   0
295    295  0.901062   0
296    296  0.902312   0
297    297  0.901637   0
298    298  0.903383   0
299    299  0.901915   0
300    300  0.896558   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
7-th fold's auc_score:0.8395127118644067
[[231   5]
 [ 20   4]]

==== Training Phase ====
8-th fold
labels:(array([-1,  1]), array([ 217, 2123]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.2057): 100%|██████████| 300/300 [08:29<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.332699   0
2        2  0.188481   0
3        3  0.176652   0
4        4  0.178119   0
5        5  0.198962   0
6        6  0.210552   0
7        7  0.219686   0
8        8  0.234021   0
9        9  0.251248   0
10      10  0.268287   0
11      11  0.284415   0
12      12  0.302019   0
13      13  0.323149   0
14      14  0.351046   0
15      15  0.380627   0
16      16  0.422472   0
17      17  0.491045   0
18      18  0.562431   0
19      19  0.622345   0
20      20  0.667103   0
21      21  0.705025   0
22      22  0.738479   0
23      23  0.762462   0
24      24  0.774192   0
25      25  0.783623   0
26      26  0.793014   0
27      27  0.803198   0
28      28  0.814689   0
29      29  0.823982   0
..     ...       ...  ..
271    271  0.893149   0
272    272  0.893921   0
273    273  0.893684   0
274    274  0.893545   0
275    275  0.893208   0
276    276  0.892257   0
277    277  0.893168   0
278    278  0.894179   0
279    279  0.893585   0
280    280  0.893505   0
281    281  0.894139   0
282    282  0.893862   0
283    283  0.893624   0
284    284  0.893367   0
285    285  0.893525   0
286    286   0.89402   0
287    287  0.893822   0
288    288  0.895189   0
289    289  0.894833   0
290    290  0.893327   0
291    291  0.892931   0
292    292   0.89402   0
293    293  0.894734   0
294    294  0.894833   0
295    295  0.894516   0
296    296  0.893545   0
297    297  0.895031   0
298    298  0.895447   0
299    299  0.895506   0
300    300  0.894337   0

[301 rows x 3 columns]

/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


==== Test Phase ====
8-th fold's auc_score:0.9188829787234043
[[234   1]
 [ 21   3]]

==== Training Phase ====
9-th fold
labels:(array([-1,  1]), array([ 217, 2123]))


Loss:   0%|          | 0/300 [00:00<?, ?it/s]


Training started.



SGCN (Loss=0.1985): 100%|██████████| 300/300 [08:24<00:00,  1.65s/it]


0         1   2
0    Epoch       AUC  F1
1        1  0.305754   0
2        2  0.232624   0
3        3  0.209978   0
4        4  0.199616   0
5        5  0.213227   0
6        6  0.226026   0
7        7  0.231574   0
8        8  0.241441   0
9        9  0.254398   0
10      10  0.265613   0
11      11  0.283583   0
12      12  0.300781   0
13      13  0.321089   0
14      14  0.347579   0
15      15  0.377833   0
16      16  0.414646   0
17      17  0.468061   0
18      18  0.533821   0
19      19  0.595162   0
20      20  0.647428   0
21      21   0.70762   0
22      22  0.759689   0
23      23  0.773409   0
24      24  0.781572   0
25      25  0.785327   0
26      26  0.786178   0
27      27  0.792221   0
28      28   0.79768   0
29      29  0.802841   0
..     ...       ...  ..
271    271  0.860794   0
272    272  0.859328   0
273    273  0.861309   0
274    274  0.860279   0
275    275  0.860021   0
276    276   0.85808   0
277    277  0.857525   0
278    278  0.860259   0
279    279  0.861032   0
280    280  0.858615   0
281    281   0.86549   0
282    282  0.865212   0
283    283  0.865589   0
284    284   0.86445   0
285    285  0.864935   0
286    286  0.863489   0
287    287   0.85489   0
288    288  0.853681   0
289    289  0.859348   0
290    290  0.863707   0
291    291  0.861329   0
292    292  0.862855   0
293    293  0.862538   0
294    294  0.861012   0
295    295  0.858971   0
296    296  0.856495   0
297    297  0.862974   0
298    298  0.867788   0
299    299  0.864063   0
300    300  0.861686   0

[301 rows x 3 columns]

==== Test Phase ====
9-th fold's auc_score:0.925709219858156
[[233   2]
 [ 21   3]]



/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/ipykernel/__main__.py:32: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


In [9]:
np.mean(auc_scores)

0.869094813078495

In [ ]:
_ = plt.hist(probabilities[:,1][test_node_labels==1],alpha=0.5,bins=10)
_ = plt.hist(probabilities[:,1][test_node_labels==-1],alpha=0.5,bins=10)
# plt.xlim(0,0.03)

# new_balance_theoryの結果


結果 (32,16) (non-class-weights,10e-3, test_size=0.33)
- amazon : 0.8650228
- alpha : 0.984
- otc : 0.9979548

結果 (32,16) (non-class-weights,10e-4, test_size=0.33)
- amazon : 0.8690948130
- alpha : 0.97264
- otc : 0.996

結果 (32,16, 8) (non-class-weights,10e-4, test_size=0.33)
- amazon : 0.8722951
- alpha : 
- otc : 

結果 (32,16) (non-class-weights,10e-3, test_size-0.5)
- amazon : 
- alpha : 
- otc : 

結果 (32,32,32,32) (non-class-weights,10e-3, hidden_residual,) (100 epoch)
- amazon : 0.86009158252
- alpha : 
- otc : 

結果 (32,32,32,) (class-weights,10e-3,hidden_residual) (100 epoch)
- amazon : 0.864321831
- alpha : 
- otc : 

---

結果 (32,) (no-class-weights, 0.33)
- amazon : 0.856184081
- alpha : 0.983
- otc : 0.998

結果 (32,16) (no-class-weights,test_size=0.33,weight_decay=10e-5)
- amazon : 0.871
- alpha : 0.9828
- otc : 0.997

結果 (32,16,8) (no-class-weights,test_size=0.33)
- amazon : 0.867
- alpha : 0.97123376
- otc : 0.9883699633699635

結果 (32,16,8) (no-class-weights,test_size=0.33,lamb=0)
- amazon : 0.871944
- alpha : 0.96454545
- otc : 0.994810744810

結果 (32,32,32) (no-class-weights,test_size=0.33,lamb=1.0,residual)
- amazon : 0.869
- alpha : 0.
- otc : 0.

結果 (32,16) (no-class-weights,test_size=0.33,weight_decay=10e-3)
- amazon : 0.871668935569  -->  0.866652196778459
- alpha : 0.9868181818181819 --> 0.9778071928071927
- otc : 0.9963980463980464  --> 0.99557387057

結果 (32,16) (no-class-weights,test_size=0.33,weight_decay=10e-3, amazon_unsigned)
- amazon : 
- alpha : 0.9868181818181819 --> 0.9778071928071927
- otc : 0.9963980463980464  --> 0.99557387057

---

結果 (32,16) (no-class-weights,test_size=0.66)
- amazon : 0.8635876758
- alpha : 0.9831
- otc : 0.993

結果 (64,32)
- amazon : 0.8597141032576031
- alpha : 
- otc : 

結果 (64,32,16,8)
- amazon : 0.8516649987979326
- alpha : 
- otc : 

結果
- amazon : 0.8704, (weighted classes loss : 0.871)
- alpha : (sampled: 0.9804), (normal: 0.9857)
- epinions : 
- otc : (0.9947), (normal:0.996)

## single-validation

In [ ]:
args = easydict.EasyDict({
        "edge_path": '../input/amazon/amazon_network.csv',#'../input/amazon/user_network.csv',
        "features_path":  '../input/amazon/amazon_node_feature.csv',#'../input/amazon/user_network.csv',
        "nodes_path": '../input/amazon/amazon_gt.csv',
        "embedding_path": '../tmp/embedding/amazon_sgcn_feature05.pkl', # tmp folder for cross-validation
        "regression_weights_path": '../tmp/weights/amazon_sgcn_feature05.pkl',
        "inductive_model_path": '../output/inductive/amazon_model', # or None
        "log_path": '../logs/amazon_logs_feature05.json',
        "epochs":500,
        "test_size":0.33,
        "reduction_iterations": 128,
        "reduction_dimensions": 30,
        "seed": 42,
        "lamb": 1.0,
        "learning_rate": 0.005,  
        "weight_decay": 10e-5, 
        # "layers": [64, 32,16,8],
        "layers": [32,16],
        "spectral_features":False,
        "general_features": True,  
        "sample_num":None,
        "class_weights":False,
        "node_under_sampling":False,
})

In [ ]:
#tab_printer(args)
edges, nodes_dict = read_graph(args) # nodes_dict['indice']:node_id , nodes_dict['label'] : label

In [ ]:
trainer = SignedGCNTrainer(args, edges, nodes_dict)
trainer.setup_dataset()
trainer.create_and_train_model()

In [ ]:
if args.test_size > 0:
    trainer.save_model()
    # score_printer(trainer.logs)
    display(pd.DataFrame(trainer.logs['performance']))
    save_logs(args, trainer.logs)

#### inductive settings

In [ ]:
new_args = easydict.EasyDict({
        "edge_path": '../input/amazon_music/amazon_music_network.csv',#'../input/amazon_music/user_network.csv',
        "features_path":  '../input/amazon_music/amazon_music_node_feature.csv',#'../input/amazon_music/user_network.csv',
        "nodes_path": '../input/amazon_music/amazon_music_gt.csv',
        "embedding_path": '../tmp/embedding/amazon_music_sgcn_feature05.csv', # tmp folder for cross-validation
        "regression_weights_path": '../tmp/weights/amazon_music_sgcn_feature05.csv',
        "inductive_model_path": None, # '../output/inductive/amazon_music_model', # or None
        "log_path": '../logs/amazon_music_logs_feature05.json',
        "epochs":300,
        "test_size":0.33,
        "reduction_iterations": 128,
        "reduction_dimensions": 30,
        "seed": 42,
        "lamb": 1.0,
        "learning_rate": 0.005,  
        "weight_decay": 10e-5, 
        # "layers": [64, 32,16,8],
        "layers": [32, 16],
        "spectral_features":False,
        "general_features": True,  
        "class_weights":False,
        "sample_num":None,
        "node_under_sampling":False,
})

In [ ]:
new_edges, new_nodes_dict = read_graph(new_args)

X = np.array(pd.read_csv('../input/amazon_elec/amazon_elec_node_feature.csv')) # general node features

In [ ]:
predictor = SignedGCNPredictor(new_args, '../output/inductive/amazon_model', X, new_edges,new_nodes_dict)

In [ ]:
predictions = predictor.predict()
predict_labels = predictions.argmax(1)

In [ ]:
y_true = new_nodes_dict['label']

In [ ]:
roc_auc_score(y_true=[1 if i==-1 else 0 for i in new_nodes_dict['label']],y_score=predictions[:,1][new_nodes_dict['indice']])

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix
confusion_matrix([1 if i==-1 else 0 for i in new_nodes_dict['label']],predict_labels[new_nodes_dict['indice']])

In [ ]:
y_true.shape

In [ ]:
_ = plt.hist(predictions[:,1][new_nodes_dict['indice']][y_true==1],alpha=0.5,bins=10)
_ = plt.hist(predictions[:,1][new_nodes_dict['indice']][y_true==-1],alpha=0.5,bins=10)

結果
- amazon
    - elec : 0.80787063118
    - movie : 0.55
    - music : 0.76
    

In [ ]:
import json

In [ ]:
performance = pd.DataFrame(json.load(open('../logs/otc_logs_feature05.json','r'))['performance'])

performance.columns = performance.iloc[0,:]

performance = performance.iloc[1:,:]

In [ ]:
performance['AUC'].plot()

In [ ]:
feature = pd.read_csv(args.embedding_path,index_col='id')

weight = pd.read_csv(args.regression_weights_path)

In [ ]:
predictions = np.dot(feature.values,weight.values.T)

In [ ]:
probabilities = torch.nn.functional.softmax(torch.from_numpy(predictions)).numpy()

In [ ]:
predict_labels = probabilities.argmax(1)

In [ ]:
roc_auc_score(y_true=[0 if i==-1 else 1 for i in nodes_dict['label']],y_score=probabilities[:,1][nodes_dict['indice']])

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix
confusion_matrix([0 if i==-1 else 1 for i in nodes_dict['label']],predict_labels[nodes_dict['indice']])

# TODO

In [ ]:
from torch_scatter import scatter_add, scatter_mean
from torch_geometric.utils import remove_self_loops, add_self_loops

In [ ]:
positive_edges = edges["positive_edges"]

In [ ]:
positive_edges = torch.from_numpy(np.array(positive_edges, dtype=np.int64).T).type(torch.long)

In [ ]:
positive_edges.shape

In [ ]:
positive_edges, _ = remove_self_loops(positive_edges, None)

In [ ]:
positive_edges.shape

In [ ]:
positive_edges = add_self_loops(positive_edges, num_nodes=nodes_dict['all_ncount'])

In [ ]:
positive_edges.shape

In [ ]:
row, col = positive_edges

In [ ]:
row,col

In [ ]:
row.shape, col.shape

In [ ]:
x = torch.from_numpy(pd.read_csv(args.features_path).values)

In [ ]:
row.shape,col.shape,x.shape

In [ ]:
x[col].shape

In [ ]:
out = scatter_mean(x[col], row, dim=0, dim_size=x.size(0))

In [ ]:
out.shape

In [ ]:
(out.numpy()==x.numpy()).all(1).sum()

In [ ]:
x.shape

In [ ]:
np.unique(row.numpy()).shape